src/Nethermind/Nethermind.Network/NetworkStorage.cs

```diff
@@ -16,52 +20,93 @@ namespace Nethermind.Network
 {
     public class NetworkStorage : INetworkStorage
     {
+        private readonly object _lock = new();
         private readonly IFullDb _fullDb;
         private readonly ILogger _logger;
+        private readonly List<NetworkNode> _nodesList = new();
+        private readonly HashSet<PublicKey> _nodePublicKeys = new();
         private long _updateCounter;
         private long _removeCounter;
+        private NetworkNode[] _nodes;
 
         public NetworkStorage(IFullDb? fullDb, ILogManager? logManager)
         {
             _fullDb = fullDb ?? throw new ArgumentNullException(nameof(fullDb));
             _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
         }
 
-        public int PersistedNodesCount
+        public int PersistedNodesCount => GetPersistedNodes().Length;
+
+        public NetworkNode[] GetPersistedNodes()
```

Comment:
Does it have to return an Array? Can it just return IReadOnlyCollection and we could leverage the list itself? We could avoid storing array and list and allocating array all the time.

Reply:
Recreating the array gets around the concurrency problem of the list being used while the list is being adjusted since its used in a few places and the nodes are created via events from the network; so it would need tighter synchronisation on the iteration in `DiscoveryApp.AddPersistedNodes`; `NodesLoader.LoadPeersFromDb` and `PeerPool.UpdateReputationAndMaxPeersCount()` with the node creation form `ProtocolsManager.InitSyncPeerProtocol`; `DiscoveryApp.RunDiscoveryCommit()`; `DiscoveryManager.GetNodeLifecycleManager` and removal from `DiscoveryManager.OnIncomingMsg` and `PeerPool.CleanupPersistedPeers`

It doesn't have the issue currently as it completely recreates the `NetworkNode[]` from scratch from the DB each time the method is called; whereas here its reused form most method calls (and it only reads from the DB once for app lifetime, though it continues to write changes to it, but updates in memory structure instead of reading back)

**tl;dr** yes it could, but is complicated due to the concurrency

Reply:
Fair enough, but you could have 2 lists that you would only copy from one to the other without any allocations or GC

Reply:
🤔 could work; though use of the other list isn't guaranteed to have finished before starting to repopulate it

However thinking about that, could maybe get away with one List by taking a bounded Span reference via [`CollectionsMarshal.AsSpan<T>(List<T>)`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan?view=net-7.0); then adding some null checks when the items are returned and skipping them. Would need to look into it a bit more as might be a bit wild 😅

Reply:
We have this two list approach in `TxBroadcaster` with `_accumulatedTemporaryTxs` and `_txsToSend`

Reply:
Looks like its a single threaded reader and when the method is re-entered you know the iteration is complete? (e.g. timer fire); whereas here its read and written from a bunch of sources?

Reply:
Tried the 2 approach + Span, but is tricky to workout when to switch without it immediately switching back (e.g. `UpdateNodes(IEnumerable<NetworkNode> nodes)` calls `UpdateNode(NetworkNode node)` multiple times); and the Span still suffers from nodes being removed in the middle of the list, so skipping nodes 🤔

Reply:
Probably better to look at in future PR?